Shopify/shopify_app
# Add GraphQL adapter for WebhooksManager

### What this PR does

This PR introduces a GraphQL adapter for webhooks. It allows switching between REST and GraphQL for webhooks creation/deletion via config setting:
```ruby
config.webhook_api_adapter = :rest # default
config.webhook_api_adapter = :graphql
```
The API for webhooks management stays the same, only communication with Shopify API changes.
```ruby
config.webhooks = [
  { topic: 'app_subscriptions/update', address: "https://#{app_host}/webhooks/app_subscriptions_update" },
  { topic: 'app/uninstalled', address: "https://#{app_host}/webhooks/app_uninstalled" },
]
```

### Why do we need GraphQL for webhooks?

Unfortunately, not all webhook topics are available in REST API. For example, the topic `app_subscriptions/update` is available only in GraphQL and according to Shopify Staff replies in community threads this not going to change anytime soon:
https://community.shopify.com/c/Shopify-APIs-SDKs/webhook-for-the-topic-quot-appsubscriptions-update-quot-failing/td-p/562173
https://community.shopify.com/c/Shopify-APIs-SDKs/app-subscription-update-not-listed-anymore/td-p/581015

Since `app_subscriptions/update` is a very important webhook for any paid app I think it's a good idea to support GraphQL webhooks in this gem.

This should also resolve this issue: #1016 

### Reviewer's guide to testing

You will need a dev shop and rails app with shopify_app installed to test it.

1. Switch to graphql adapter and register `app_subscriptions/update` webhook.
2. Trigger webhook creation via following code:
```ruby
Shop.find_each do |shop|
  webhooks = ShopifyApp.configuration.webhooks
  adapter = ShopifyApp.configuration.webhook_api_adapter

  shop.with_shopify_session do
    manager = ShopifyApp::WebhooksManager.new(webhooks, adapter: adapter)
    manager.create_webhooks
  end
end
```
3. Check in graphql that webhook was added.
4. Trigger webhook destroy via following code:
```ruby
Shop.find_each do |shop|
  webhooks = ShopifyApp.configuration.webhooks
  adapter = ShopifyApp.configuration.webhook_api_adapter

  shop.with_shopify_session do
    manager = ShopifyApp::WebhooksManager.new(webhooks, adapter: adapter)
    manager.destroy_webhooks
  end
end
```
5. Check in graphql that webhook was remove.

I'm using this code in production already and it works smoothly.

### Checklist

Before submitting the PR, please consider if any of the following are needed:
